python/ray/util/state/state_manager.py

```diff
@@ -292,40 +314,104 @@ async def get_all_task_info(
+                    filter_predicate,
+                    GetTaskEventsRequest.Filters.ActorIdFilter(),
+                    req_filters.actor_filters,
+                    "actor_id",
+                )
+
+                if new_actor_id:
+                    actor_id = new_actor_id
+
             elif key == "job_id":
-                req_filters.job_filter.job_id = JobID(hex_to_binary(value)).binary()
-                req_filters.job_filter.predicate = filterPredicate
+                is_empty_reply, new_job_id = self.add_task_filter(
+                    job_id,
+                    JobID(hex_to_binary(value)).binary(),
+                    filter_predicate,
+                    GetTaskEventsRequest.Filters.JobIdFilter(),
+                    req_filters.job_filters,
+                    "job_id",
+                )
+
+                if new_job_id:
+                    job_id = new_job_id
+
             elif key == "task_id":
-                task_id_filter = GetTaskEventsRequest.Filters.TaskIdFilter()
-                task_id_filter.task_id = TaskID(hex_to_binary(value)).binary()
-                task_id_filter.predicate = filterPredicate
-                req_filters.task_filters.append(task_id_filter)
+                is_empty_reply, new_task_id = self.add_task_filter(
+                    task_id,
+                    TaskID(hex_to_binary(value)).binary(),
+                    filter_predicate,
+                    GetTaskEventsRequest.Filters.TaskIdFilter(),
+                    req_filters.task_filters,
+                    "task_id",
+                )
+
+                if new_task_id:
+                    task_id = new_task_id
+
             elif key == "name":
-                req_filters.name_filter.name = value
-                req_filters.name_filter.predicate = filterPredicate
+                is_empty_reply, new_task_name = self.add_task_filter(
+                    task_name,
+                    value,
+                    filter_predicate,
+                    GetTaskEventsRequest.Filters.TaskNameFilter(),
+                    req_filters.task_name_filters,
+                    "task_name",
+                )
+
+                if new_task_name:
+                    task_name = new_task_name
+
             elif key == "state":
-                req_filters.state_filter.state = value
-                req_filters.state_filter.predicate = filterPredicate
+                is_empty_reply, new_state = self.add_task_filter(
+                    state,
+                    value,
+                    filter_predicate,
+                    GetTaskEventsRequest.Filters.StateFilter(),
+                    req_filters.state_filters,
+                    "state",
+                )
+
+                if new_state:
+                    state = new_state
+
             else:
                 continue
 
+            if is_empty_reply:
+                # return empty reply when more than 2 equal filters with the same filter
```

Comment:
hmm, i wonder if we could do the checking earlier than here? Instead of checking if early returns are possible while building the actual protobuf filter, could we do the checking on `filters` object?

Reply:
Good point! If all filters are with AND semantics, we should be able to move the early return to a more general location. Let me check it out.